zerver/tests/test_timeout.py

```diff
@@ -0,0 +1,53 @@
+import time
+import traceback
+
+from zerver.lib.test_classes import ZulipTestCase
+from zerver.lib.timeout import TimeoutExpired, timeout
+
+
+class TimeoutTestCase(ZulipTestCase):
+    # We can't use assertRaises because that doesn't store the
+    # traceback, which we want to verify
+
+    def something_exceptional(self) -> int:
+        raise ValueError("Something went wrong")
+
+    def sleep_x_seconds_y_times(self, x: float, y: int) -> int:
+        for i in range(y):
+            time.sleep(x)
+        return 42  # nocoverage
+
+    def test_timeout_returns(self) -> None:
+        ret = timeout(1, lambda: 42)
+        self.assertEqual(ret, 42)
+
+    def test_timeout_exceeded(self) -> None:
+        try:
+            timeout(1, lambda: self.sleep_x_seconds_y_times(0.1, 50))
+            self.assertTrue(False, "Failed to raise a timeout")  # nocoverage
```

Comment:
This might be better done with `raise AssertionError`, which comes with an automatic `nocoverage` and also would make clear this isn't intended to ever succeed.

Reply:
Ah!  Great solution, and good to know about.